### PR TITLE
Feature/ccpp constraint analysis (Old repo)

### PR DESCRIPTION
**Pull request by fabianbs96**
The implementation of ConstraintAnalysis is now done. Please review the code before merging

### CODE REVIEW

File: CogniCrypt++/lib/CrySL/FormatConverter/CallToAnalysis/IDECallToAnalysis.cpp [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+// implement the flow/edge functions
+std::shared_ptr<psr::FlowFunction<IDECallToAnalysis::d_t>>
+IDECallToAnalysis::getNormalFlowFunction(IDECallToAnalysis::n_t curr,
+                                         IDECallToAnalysis::n_t succ) {
+
+  if (auto load = llvm::dyn_cast<llvm::LoadInst>(curr)) {
+    return std::make_shared<psr::LambdaFlow<d_t>>(
+        [=](d_t source) -> std::set<d_t> {
+          if (source.getStatefulObject() == load->getPointerOperand()) {
+            // std::cout << "Load " << FlowDomain(source.getEvent(), load)
```
Comment (sammekh) [inline CogniCrypt++/lib/CrySL/FormatConverter/CallToAnalysis/IDECallToAnalysis.cpp]: remove the commented std::couts if no longer required later

File: CogniCrypt++/include/CCPP/CrySL/Parser/CrySLParser.h [hunk 2]
```diff
@@ -114,623 +209,612 @@ class  CrySLParser : public antlr4::Parser {
...
+  SuParListContext *suParList();
 
-  class  SuParContext : public antlr4::ParserRuleContext {
+  class SuParContext : public antlr4::ParserRuleContext {
   public:
-    CrySLParser::ConsPredContext *value = nullptr;;
-    antlr4::Token *thisptr = nullptr;;
-    antlr4::Token *wildcard = nullptr;;
+    CrySLParser::ConsPredContext *value = nullptr;
+    ;
```
Comment (ahmedammad) [inline CogniCrypt++/include/CCPP/CrySL/Parser/CrySLParser.h]: typo maybe  ?

File: CogniCrypt++/include/CCPP/CrySL/Parser/CrySLParser.h [hunk 3]
```diff
@@ -114,623 +209,612 @@ class  CrySLParser : public antlr4::Parser {
...
+    antlr4::Token *thisptr = nullptr;
+    ;
+    antlr4::Token *wildcard = nullptr;
+    ;
```
Comment (ahmedammad) [inline CogniCrypt++/include/CCPP/CrySL/Parser/CrySLParser.h]: typo here and in above line

File: CogniCrypt++/include/CCPP/CrySL/Parser/CrySLParser.h [hunk 4]
```diff
@@ -114,623 +209,612 @@ class  CrySLParser : public antlr4::Parser {
...
-    antlr4::Token *exp = nullptr;;
-    antlr4::Token *nullLit = nullptr;;
+    antlr4::Token *base = nullptr;
+    ;
+    antlr4::Token *pow = nullptr;
+    ;
+    antlr4::Token *exp = nullptr;
+    ;
+    antlr4::Token *nullLit = nullptr;
+    ;
```
Comment (ahmedammad) [inline CogniCrypt++/include/CCPP/CrySL/Parser/CrySLParser.h]: extra semicolon ? two more occurrances

File: CogniCrypt++/lib/CrySL/FormatConverter/CallToAnalysis/IDECallToAnalysis.cpp [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+std::shared_ptr<psr::FlowFunction<IDECallToAnalysis::d_t>>
+IDECallToAnalysis::getCallToRetFlowFunction(n_t callSite, n_t retSite,
+                                            std::set<m_t> callees) {
+  llvm::ImmutableCallSite CS(callSite);
+  auto evt = OrderTypeStateDescription::getEventWithParams(
+      psr::cxx_demangle(CS.getCalledFunction()->getName()), CS, mapFuncEvents,
+      lcp);
+  if (evt) {
+
+    // auto ret = std::make_shared<psr::Gen<d_t>>(
```
Comment (sammekh) [inline CogniCrypt++/lib/CrySL/FormatConverter/CallToAnalysis/IDECallToAnalysis.cpp]: this commented code is not being used so you can remove it

File: CogniCrypt++/lib/CrySL/FormatConverter/CallToAnalysis/IDECallToAnalysis.cpp [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+          }
+        }
+        os << '\n';
+      }
+    }
+    os << '\n';
+  }
+}
+
+void IDECallToAnalysis::stripBottomResults(
```
Comment (sammekh) [inline CogniCrypt++/lib/CrySL/FormatConverter/CallToAnalysis/IDECallToAnalysis.cpp]: will this be implemented?

File: CogniCrypt++/lib/CrySL/FormatConverter/CallToAnalysis/MapObjectToValue.cpp [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+      // auto litVal = getConstant(cs.getInstruction(), cs.getArgOperand(i));
+      auto litVal =
+          getConstantValue(lcp, cs.getInstruction(), cs.getArgOperand(i));
+      auto singleLitVal =
+          litVal.size() == 1 ? *litVal.begin() : LCUtils::EdgeValue(nullptr);
+      if (!singleLitVal.isTop()) {
+        objectOwner.emplace_back(new ObjectLLVM(singleLitVal, ctx));
+        map[objEx->getObject().get()] = objectOwner.back().get();
+        // std::cout << "\tObject " << objEx->getObject()->getName()
```
Comment (sammekh) [inline CogniCrypt++/lib/CrySL/FormatConverter/CallToAnalysis/MapObjectToValue.cpp]: remove debug output

File: CogniCrypt++/lib/CrySL/FormatConverter/CallToAnalysis/FlowDomain.cpp [hunk 8]
```diff
@@ -0,0 +1,47 @@
+#include <CCPP/CrySL/FormatConverter/CallToAnalysis/FlowDomain.h>
+#include <phasar/Utils/LLVMShorthands.h>
+
+namespace CCPP::CallToAnalysisUtils {
+FlowDomain::FlowDomain(std::nullptr_t) : evt(0), statefulObject(0) {}
+FlowDomain::FlowDomain(const Event *evt, const llvm::Value *obj)
+    : evt(evt), statefulObject(obj) {
+  // if (!statefulObject) {
```
Comment (ahmedammad) [inline CogniCrypt++/lib/CrySL/FormatConverter/CallToAnalysis/FlowDomain.cpp]: remove commented code if it is not required anymore

File: CogniCrypt++/lib/CrySL/FormatConverter/ObjectToValueAnalysis/IDEObjectToValueAnalysis.cpp [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+            /*retSiteNode == statefulObj*/ aliases.count(statefulObj) ||
+        retSiteNode.isEmpty() && !statefulObj) {
+      ObjectToValueDomain ret;
+      // std::cout << "At " << evt->getEventName() << ": " << std::endl;
+
+      mapObjectToValue(ret, cs, evt, lcp, objectOwner, llvmCtx);
+      // Utils::print(std::cout << "Generate for " << retSiteNode, ret);
+      return std::make_shared<GenerateEdgeFunction>(ret);
+    }
+    // else {
```
Comment (sammekh) [inline CogniCrypt++/lib/CrySL/FormatConverter/ObjectToValueAnalysis/IDEObjectToValueAnalysis.cpp]: is there's no case where this if won't be false?

File: CogniCrypt++/lib/CrySL/FormatConverter/ObjectToValueAnalysis/JoinEdgeFunction.cpp [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    return true;
+  if (auto otherJoin = dynamic_cast<JoinEdgeFunction *>(other.get())) {
+    return (frst->equal_to(otherJoin->frst) &&
+            scnd->equal_to(otherJoin->scnd)) // join is commutative...
+           ||
+           (frst->equal_to(otherJoin->scnd) && scnd->equal_to(otherJoin->frst));
+  }
+  return false;
+}
+/*void JoinEdgeFunction::print(std::ostream &OS, bool isForDebug) const {
```
Comment (sammekh) [inline CogniCrypt++/lib/CrySL/FormatConverter/ObjectToValueAnalysis/JoinEdgeFunction.cpp]: remove if not being used

**Pull request merged**